pkg/controllers/machine/termination_test.go

```diff
@@ -42,34 +65,373 @@ var _ = Describe("Termination", func() {
 			},
+			Status: v1alpha5.MachineStatus{
+				ProviderID: test.RandomProviderID(),
+				Capacity: v1.ResourceList{
+					v1.ResourceCPU:    resource.MustParse("10"),
+					v1.ResourceMemory: resource.MustParse("100Mi"),
+					v1.ResourcePods:   resource.MustParse("110"),
+				},
+				Allocatable: v1.ResourceList{
+					v1.ResourceCPU:    resource.MustParse("8"),
+					v1.ResourceMemory: resource.MustParse("80Mi"),
+					v1.ResourcePods:   resource.MustParse("110"),
+				},
+			},
+		})
+		node.Finalizers = []string{v1alpha5.TerminationFinalizer}
+	})
+	It("should cordon, drain, and delete the Machine on terminate", func() {
+		ExpectApplied(ctx, env.Client, provisioner, node, machine)
+
+		// Kickoff the deletion flow for the machine
+		Expect(env.Client.Delete(ctx, machine)).To(Succeed())
+
+		// Machine should delete and the Node deletion should cascade shortly after
+		ExpectReconcileSucceeded(ctx, machineController, client.ObjectKeyFromObject(machine))
+		ExpectNotFound(ctx, env.Client, machine)
+	})
+	It("should not race if deleting machines in parallel", func() {
+		var machines []*v1alpha5.Machine
+		for i := 0; i < 10; i++ {
+			machine, node = test.MachineAndNode(v1alpha5.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Finalizers: []string{
+						v1alpha5.TerminationFinalizer,
+					},
+					Labels: map[string]string{
+						v1alpha5.ProvisionerNameLabelKey: provisioner.Name,
+					},
+				},
+				Spec: v1alpha5.MachineSpec{
+					Resources: v1alpha5.ResourceRequirements{
+						Requests: v1.ResourceList{
+							v1.ResourceCPU:          resource.MustParse("2"),
+							v1.ResourceMemory:       resource.MustParse("50Mi"),
+							v1.ResourcePods:         resource.MustParse("5"),
+							fake.ResourceGPUVendorA: resource.MustParse("1"),
+						},
+					},
+				},
+				Status: v1alpha5.MachineStatus{
+					ProviderID: test.RandomProviderID(),
+					Capacity: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse("10"),
+						v1.ResourceMemory: resource.MustParse("100Mi"),
+						v1.ResourcePods:   resource.MustParse("110"),
+					},
+					Allocatable: v1.ResourceList{
+						v1.ResourceCPU:    resource.MustParse("8"),
+						v1.ResourceMemory: resource.MustParse("80Mi"),
+						v1.ResourcePods:   resource.MustParse("110"),
+					},
+				},
+			})
+			ExpectApplied(ctx, env.Client, machine, node)
+			Expect(env.Client.Delete(ctx, machine)).To(Succeed())
+			machines = append(machines, machine)
+		}
+
+		var wg sync.WaitGroup
+		// this is enough to trip the race detector
+		for i := 0; i < 10; i++ {
+			wg.Add(1)
+			go func(machine *v1alpha5.Machine) {
+				defer GinkgoRecover()
+				defer wg.Done()
+				ExpectReconcileSucceeded(ctx, machineController, client.ObjectKeyFromObject(machine))
+			}(machines[i])
+		}
+		wg.Wait()
+		ExpectNotFound(ctx, env.Client, lo.Map(machines, func(m *v1alpha5.Machine, _ int) client.Object { return m })...)
+	})
+	It("should exclude machines from load balancers when terminating", func() {
```

Comment:
Is this a new feature? I honestly don't know what this refers to.

Reply:
Added a comment that gives a brief summary and references the issue